src/data/set/intervals/monotone.lean

```diff
@@ -177,3 +178,84 @@ calc (⋃ x, Ioo (f x) (g x)) = (⋃ x, Ioi (f x)) ∩ ⋃ x, Iio (g x) :
 ... = Ioi a ∩ Iio b : congr_arg2 (∩) ha.Union_Ioi_eq hb.Union_Iio_eq
 
 end Union
+
+section succ_order
+
+open order
+
+variables {α β : Type*} [partial_order α] [succ_order α] [is_succ_archimedean α]
+
+lemma strict_mono_on.Iic_id_le [order_bot α] {n : α} {φ : α → α}
+  (hφ : strict_mono_on φ (set.Iic n)) :
+  ∀ m ≤ n, m ≤ φ m :=
+begin
+  revert hφ,
+  refine succ.rec_bot (λ n, strict_mono_on φ (set.Iic n) → ∀ m, m ≤ n → m ≤ φ m) _ _ _,
+  { simp_rw le_bot_iff,
+    rintro _ m rfl,
+    exact bot_le },
+  { rintro k ih hφ m hm,
+    by_cases hk : is_max k,
+    { rw succ_eq_iff_is_max.2 hk at hm,
+      exact ih (hφ.mono $ Iic_subset_Iic.2 (le_succ _)) _ hm },
+    obtain (rfl | h) := le_succ_iff_eq_or_le.1 hm,
+    { specialize ih (strict_mono_on.mono hφ (λ x hx, le_trans hx (le_succ _))) k le_rfl,
+      refine le_trans (succ_mono ih) (succ_le_of_lt (hφ (le_succ _) le_rfl _)),
+      rw lt_succ_iff_eq_or_lt_of_not_is_max hk,
+      exact or.inl rfl },
+    { exact ih (strict_mono_on.mono hφ (λ x hx, le_trans hx (le_succ _))) _ h } }
+end
+
+variables [preorder β] {ψ : α → β}
+
+lemma strict_mono_on_Iic_of_lt_succ [order_bot α]
```

Comment:
I don't understand why you need `order_bot` here. Also, wouldn't it be more natural to formulate your assumption with `m < n`?

Reply:
I think we need `order_bot` if we don't assume `is_pred_achimedian` since otherwise there is no guarantee the reverse induction will reach every element less than `n`.

Reply:
You are supposing archimedean, so you can do the induction in the other direction: if you have `m' < m`, you can start your induction from `m'` all the way to `m`, no?

ok for `succ m \le n`, I had overlooked the subtlety. It's maybe worth mentioning in the docstring.

Reply:
Thanks, I did not realize I had already assumed `is_succ_archimedean`. I have removed the `order_bot` assumption now.